library/PreseedControl.py

```diff
@@ -0,0 +1,117 @@
+
+        try:
+            preseeds = self.requester.get(url)
+        except Exception as err:
+            raise err
+
+        for preseed in preseeds:
+            if preseed['name'] == self.name:
+                return preseed
+
+        return False
+
+    def get_preseed_id(self):
+        try:
+            preseed = self.check_for_existence()
+        except Exception as err:
+            raise err
+
+        if preseed != False:
+            return preseed['id']
+        else:
+            return None
+
+    def upload_preseed(self, preseed_file, preseed_type, preseed_desc=None,
+                       public=False, knowngood=False):
+        """Uploads preseed. Should check first that preseed doesn't exist, else
+        it modifies preseed (separate function ?). Post to upload, Put to
+        modify (+ id). Maybe implement a jinja2 syntax check ? But that should
+        be done on mrp's side"""
+
+        preseed_id = None
+        preseed_id = self.get_preseed_id()
+
+        if preseed_id is None and preseed_file == '':
+            raise ProvisionerError('Preseed does not exist and file not given')
+
+        if preseed_id is not None and preseed_file != '':  # Exists and file given
+            try:
+                return self._modify_preseed(preseed_id, 'PUT', preseed_file,
+                                           preseed_type, preseed_desc, public,
+                                           knowngood)
+            except ProvisionerError as err:
+                raise err
+
+        elif preseed_id is None and preseed_file != '':  # Doesn't exist and file given
+            try:
+                return self._modify_preseed(preseed_id, 'POST', preseed_file,
+                                           preseed_type, preseed_desc, public,
+                                          knowngood)
+            except ProvisionerError as err:
+                raise err
+
+        else:  # Exists and file not given, is it useful fetching contents?
+            return "Preseed exists"
+
+    def _modify_preseed(self, preseed_id, method, preseed_file,
+                        preseed_type, preseed_desc=None, public=False,
+                        knowngood=False):
+        preseed = self._get_preseed_from_file(preseed_file, preseed_type, preseed_desc,
+                                              public, knowngood)
+        url = '/api/v1/preseed'
+        if method == 'PUT':
+            if preseed_id is None:
+                raise ProvisionerError(
+                    'preseed ID is undefined, please use upload_preseed')
+            url = url + '/' + str(preseed_id)
+
+            try:
+                return self.requester.put(url, data=json.dumps(preseed))
+            except Exception as err:
+                raise err
+
+        elif method == 'POST':
+            try:
+                return self.requester.post(url, data=json.dumps(preseed))
+            except Exception as err:
+                raise err
```

Comment:
else: throw ProvisionError("invalid method") at least. But I don't like the idea of an external function controlling the method. Nor I think it's necessary that we have this, if we make the function above only run once.

Following my comment above to change the method via conditions, just move this code up.